[aes, dv] Debug and enable stress_all test

### Description

After discussing with @rasmus-madsen it should be noted that for AES, we have the aes_stress_vseq which covers most of the functionality that the stress_all and stress_all_with_rand_reset tests cover. The risk that we miss something for M2.5.2 by not having these two tests is very very low. That's also why we haven't bothered to enable them earlier.

For AES, the stress_all and stress_all_with_rand_reset tests are currently disabled. When trying to enable them (see [this branch on my fork](https://github.com/vogelpi/opentitan/tree/aes-dv-stress-all)), I noticed that whenever stress_all decides to reset AES in between two sequences, the simulation completely stops. It's not that the DUT doesn't react anymore or that the clock stops, but rather that the timing engine stops completely. At this point, I don't know why but I have observed this with multiple simulation tools.

To reproduce, 
1. Check out my development branch: https://github.com/vogelpi/opentitan/tree/aes-dv-stress-all
2. Run
```
util/dvsim/dvsim.py hw/ip/aes/dv/aes_unmasked_sim_cfg.hjson -t xcelium -i aes_stress_all --purge --reseed 1 --seed 2268389051 --waves shm --verbosity m
```
3. After terminating the simulation, the log should show something like:
```
UVM_INFO @  45822265 ps: (aes_stress_all_vseq.sv:40) uvm_test_top.env.virtual_sequencer [uvm_test_top.env.virtual_sequencer.aes_stress_all_vseq] iteration[2]: starting aes_reseed_vseq, resetting 1
UVM_INFO @  45822265 ps: (cip_base_vseq.sv:114) uvm_test_top.env.virtual_sequencer [uvm_test_top.env.virtual_sequencer.aes_reseed_vseq] PRE_START START
UVM_INFO @  45822265 ps: (dv_base_vseq.sv:62) uvm_test_top.env.virtual_sequencer [uvm_test_top.env.virtual_sequencer.aes_reseed_vseq] SUPER.PRE_START START
UVM_INFO @  45822265 ps: (dv_base_vseq.sv:66) uvm_test_top.env.virtual_sequencer [uvm_test_top.env.virtual_sequencer.aes_reseed_vseq] DUT_INIT START
UVM_INFO @  45822265 ps: (aes_base_vseq.sv:34) uvm_test_top.env.virtual_sequencer [uvm_test_top.env.virtual_sequencer.aes_reseed_vseq] AES DUT_INIT
UVM_INFO @  45822265 ps: (dv_base_vseq.sv:87) uvm_test_top.env.virtual_sequencer [uvm_test_top.env.virtual_sequencer.aes_reseed_vseq] DV DUT_INIT
UVM_INFO @  45822265 ps: (cip_base_vseq.sv:146) uvm_test_top.env.virtual_sequencer [uvm_test_top.env.virtual_sequencer.aes_reseed_vseq] CIP APPLY_RESET
UVM_INFO @  45822265 ps: (clk_rst_if.sv:267) [tb.edn_clk_rst_if.apply_reset] CLK_RST_IF APPLY_RESET 0 99 0 1
UVM_INFO @  45822265 ps: (clk_rst_if.sv:115) [tb.edn_clk_rst_if.wait_clks] Waiting 0 clks
UVM_INFO @  45822265 ps: (clk_rst_if.sv:279) [tb.edn_clk_rst_if.apply_reset.unmblk1] Waiting 1047 ps
UVM_INFO @  45822265 ps: (dv_base_vseq.sv:99) uvm_test_top.env.virtual_sequencer [uvm_test_top.env.virtual_sequencer.aes_reseed_vseq] DV APPLY_RESET - kind HARD
UVM_INFO @  45822265 ps: (dv_base_vseq.sv:106) uvm_test_top.env.virtual_sequencer [uvm_test_top.env.virtual_sequencer.aes_reseed_vseq] ral_name aes_reg_block
UVM_INFO @  45822265 ps: (clk_rst_if.sv:267) [tb.clk_rst_if.apply_reset] CLK_RST_IF APPLY_RESET 0 87 0 1
UVM_INFO @  45822265 ps: (clk_rst_if.sv:115) [tb.clk_rst_if.wait_clks] Waiting 0 clks
UVM_INFO @  45822265 ps: (clk_rst_if.sv:279) [tb.clk_rst_if.apply_reset.unmblk1] Waiting 3872 ps
UVM_INFO @  45823312 ps: (clk_rst_if.sv:281) [tb.edn_clk_rst_if.apply_reset.unmblk1] PULLING RESET
UVM_INFO @  45823312 ps: (clk_rst_if.sv:115) [tb.edn_clk_rst_if.wait_clks] Waiting 99 clks
UVM_INFO @  45823312 ps: (clk_rst_if.sv:117) [tb.edn_clk_rst_if.wait_clks] 1 clk
UVM_INFO @  45826137 ps: (clk_rst_if.sv:281) [tb.clk_rst_if.apply_reset.unmblk1] PULLING RESET
UVM_INFO @  45826137 ps: (clk_rst_if.sv:115) [tb.clk_rst_if.wait_clks] Waiting 87 clks
UVM_INFO @  45826137 ps: (clk_rst_if.sv:117) [tb.clk_rst_if.wait_clks] 1 clk
xcelium> xmsim: *W,NCTERM: Simulation received SIGTERM signal from process 406891, user id 1000 (make).
TOOL:	xrun(64)	22.09-s001: Exiting on Jun 22, 2023 at 21:28:01 CEST  (total: 00:01:02)
make: *** [/home/pirmin/ot/opentitan/hw/dv/tools/dvsim/sim.mk:145: simulate] Error 15
```

The `apply_reset()` function for the clk_rst_if looks as follows:
```sv
...
        RstAssertAsyncDeassertSync: begin
          `dv_info($sformatf("Waiting %0d ps", dly_ps))
          #(dly_ps * 1ps);
          `dv_info("PULLING RESET")
          o_rst_n <= 1'b0;
          wait_clks(reset_width_clks);
          `dv_info("DEASSERTING RESET")
          o_rst_n <= 1'b1;
        end
...
```
And the `wait_clks()` function
```sv
 task automatic wait_clks(int num_clks);
    `dv_info($sformatf("Waiting %0d clks", num_clks))
    repeat (num_clks) begin
      `dv_info($sformatf("1 clk"))
      @cb;
    end
  endtask
```
So the `@cb` event isn't happening anymore after pulling the reset. The waves recording also stops at that point.

As mentioned above, this isn't ultra critical for M2.5.2 but we should resolve it for V3.

--- a/hw/ip/aes/data/aes_testplan.hjson
+++ b/hw/ip/aes/data/aes_testplan.hjson
@@ -7,6 +7,7 @@
                      "hw/dv/tools/dvsim/testplans/shadow_reg_errors_testplan.hjson",
                      "hw/dv/tools/dvsim/testplans/alert_test_testplan.hjson",
                      "hw/dv/tools/dvsim/testplans/tl_device_access_types_testplan.hjson",
+                     "hw/dv/tools/dvsim/testplans/stress_all_with_reset_testplan.hjson",
                      "aes_sec_cm_testplan.hjson"]
   testpoints: [
    // {
@@ -123,6 +124,14 @@
             Make sure that there is no residual data from latest operation. '''
       stage: V2
       tests: ["aes_deinit"]
+    }
+    {
+      name: stress_all
+      desc: '''
+            Run assorted sequences back-to-back.
+            '''
+      stage: V2
+      tests: ["aes_stress_all"]
     }
     {
       name: reseeding

--- a/hw/ip/aes/dv/aes_base_sim_cfg.hjson
+++ b/hw/ip/aes/dv/aes_base_sim_cfg.hjson
@@ -33,10 +33,8 @@
                 "{proj_root}/hw/dv/tools/dvsim/tests/csr_tests.hjson",
                 "{proj_root}/hw/dv/tools/dvsim/tests/shadow_reg_errors_tests.hjson",
                 "{proj_root}/hw/dv/tools/dvsim/tests/tl_access_tests.hjson",
-                "{proj_root}/hw/dv/tools/dvsim/tests/sec_cm_tests.hjson"]
-                // TODO(#19025): Debug and enable stress_all test. Note that aes_stress
-                // covers most the functionality of stress_all already.
-                // "{proj_root}/hw/dv/tools/dvsim/tests/stress_tests.hjson"]
+                "{proj_root}/hw/dv/tools/dvsim/tests/sec_cm_tests.hjson",
+                "{proj_root}/hw/dv/tools/dvsim/tests/stress_tests.hjson"]
 
   // Add additional tops for simulation.
   sim_tops: [ "aes_bind",
@@ -201,6 +199,18 @@
       uvm_test: aes_fi_test
       uvm_test_seq: aes_core_fi_vseq
       reseed: 70 // Test must cover 35 bins. Overseed by 2x to get reasonable coverage.
+    }
+    // This test runs several sequences back-to-back with random resets between the individual
+    // sequences. We shouldn't need many seeds here because aes_stress is the real stress test.
+    {
+      name: "aes_stress_all"
+      uvm_test_seq: "aes_stress_all_vseq"
+      reseed: 10
+    }
+    // A combination of aes_stress_all with random resets injected into the actual sequences.
+    {
+      name: "aes_stress_all_with_rand_reset"
+      reseed: 10
     }
   ]
 

--- a/hw/ip/aes/dv/env/aes_env.core
+++ b/hw/ip/aes/dv/env/aes_env.core
@@ -43,6 +43,7 @@ filesets:
       - seq_lib/aes_ctr_fi_vseq.sv: {is_include_file: true}
       - seq_lib/aes_core_fi_vseq.sv: {is_include_file: true}
       - seq_lib/aes_readability_vseq.sv: {is_include_file: true}
+      - seq_lib/aes_stress_all_vseq.sv: {is_include_file: true}
     file_type: systemVerilogSource
 
 generate:

--- a/hw/ip/aes/dv/env/aes_env_cfg.sv
+++ b/hw/ip/aes/dv/env/aes_env_cfg.sv
@@ -139,6 +139,9 @@ class aes_env_cfg extends cip_base_env_cfg #(.RAL_T(aes_reg_block));
   // - number of packets that were split
   int                split_cnt                  = 0;
 
+  // set when test is trying to end to indicate the last message is finished
+  bit                finish_message             = 0;
+
   // rand variables
   // 0: C model 1: OPEN_SSL/BORING_SSL
   rand bit           ref_model;

--- a/hw/ip/aes/dv/env/aes_env_cfg.sv
+++ b/hw/ip/aes/dv/env/aes_env_cfg.sv
@@ -104,11 +104,6 @@ class aes_env_cfg extends cip_base_env_cfg #(.RAL_T(aes_reg_block));
 
   rand alert_reset_trigger_e alert_reset_trigger;
 
-  // number of messages that was selected to be corrupt
-  // these should be excluded from the num_messages count
-  // when checking that all messages was processed
-  int                num_corrupt_messages       = 0;
-
   // clear register percentage
   // percentage of items that will try to clear
   // one or more registers
@@ -128,7 +123,20 @@ class aes_env_cfg extends cip_base_env_cfg #(.RAL_T(aes_reg_block));
   // dont touch updated by env //
   ///////////////////////////////
 
-  // keep track of how many packets was split
+  // keep track of:
+  // - total number of expected messages
+  int                num_messages_tot           = 0;
+  // - number of messages that was selected to be corrupt
+  //   these should be excluded from the num_messages count
+  //   when checking that all messages was processed
+  int                num_corrupt_messages       = 0;
+  // - number of good messages
+  int                good_cnt                   = 0;
+  // - number of aes_mode errors seen
+  int                corrupt_cnt                = 0;
+  // - number of skipped messages
+  int                skipped_cnt                = 0;
+  // - number of packets that were split
   int                split_cnt                  = 0;
 
   // rand variables

--- a/hw/ip/aes/dv/env/aes_scoreboard.sv
+++ b/hw/ip/aes/dv/env/aes_scoreboard.sv
@@ -20,8 +20,7 @@ class aes_scoreboard extends cip_base_scoreboard #(
   aes_seq_item complete_item;                 // merge of input and output items
   aes_seq_item key_item;                      // sequence item holding last sideload valid key
   bit          ok_to_fwd          = 0;        // 0: item is not ready to forward
-  bit          finish_message     = 0;        // set when test is trying to end
-  bit          reset_compare      = 0;        // reset compare task
+  bit          reset_rebuilding   = 0;        // reset message rebuilding task
   bit          exp_clear          = 0;        // if using sideload - we are expecting a clear
   keymgr_pkg::hw_key_req_t sideload_key = 0;  // will hold the key from sideload
   uvm_tlm_analysis_fifo #(key_sideload_item)  key_manager_fifo;
@@ -452,48 +451,71 @@ class aes_scoreboard extends cip_base_scoreboard #(
 
   // takes items from the item queue and builds full
   // aes_messages with both input data and output data.
-virtual task rebuild_message();
-    typedef enum { MSG_START,MSG_RUN } aes_message_stat_t;
-
-    aes_message_item   message, msg_clone;
-    aes_seq_item       full_item;
-    aes_message_stat_t msg_state = MSG_START;
-
+  virtual task rebuild_message();
+    typedef enum {
+      MSG_IDLE,
+      MSG_START,
+      MSG_RUN
+    } aes_message_stat_t;
+    aes_message_stat_t msg_state = MSG_IDLE;
+    aes_message_item message, msg_clone;
+    aes_seq_item full_item;
+    string txt =  "";
 
     message = new();
 
     fork
-      begin
+      begin : rebuild_messages
         forever begin
           item_fifo.get(full_item);
+          if (msg_state == MSG_IDLE) begin
+            // We have just received the very first item after a reset and can now start the regular
+            // processing.
+            msg_state = MSG_START;
+          end
           case (msg_state)
             MSG_START: begin
-              `uvm_info(`gfn, $sformatf("\t ----| got item from item fifo"), UVM_MEDIUM)
               if (!full_item.message_start()) begin
-                // check if start trigger was fired prematurely
+                // Check if e.g. the start trigger got fired prematurely and skip this message if
+                // needed.
                 if (full_item.start_item && full_item.manual_op) begin
+                  `uvm_info(`gfn, "setting skip_msg", UVM_MEDIUM)
                   message.skip_msg = 1;
-                  `uvm_info(`gfn, $sformatf("\n setting skip msg"), UVM_MEDIUM)
                 end else begin
                   `uvm_fatal(`gfn,
                       $sformatf("\n\t ----| FIRST ITEM DID NOT HAVE MESSAGE START/CONFIG SETTINGS"))
                 end
               end
+              `uvm_info(`gfn, $sformatf("rebuilding %s message, adding start item",
+                  full_item.mode.name()), UVM_MEDIUM)
               message.add_start_msg_item(full_item);
               msg_state = MSG_RUN;
             end
 
             MSG_RUN: begin
               if (full_item.message_start() || (full_item.start_item && full_item.manual_op)) begin
+                // The current item marks the start of a new message. End the previous message and
+                // add it to the message FIFO for scoring.
                 `downcast(msg_clone, message.clone());
+                `uvm_info(`gfn, $sformatf("adding %s message item of size %0d to msg_fifo",
+                    msg_clone.aes_mode.name(), msg_clone.output_msg.size()), UVM_MEDIUM)
                 msg_fifo.put(msg_clone);
                 message = new();
                 if (full_item.start_item && full_item.manual_op) begin
-                  // only set skip if this is not a real start message
-                  message.skip_msg = ~full_item.message_start() || ~full_item.data_in_valid();
+                  // Skip the message if this is item not really marks the start of a message.
+                  if (!full_item.message_start() || !full_item.data_in_valid()) begin
+                    `uvm_info(`gfn, $sformatf("setting skip_msg"), UVM_MEDIUM)
+                    message.skip_msg = 1;
+                  end else begin
+                    message.skip_msg = 0;
+                  end
                 end
+                `uvm_info(`gfn, $sformatf("rebuilding %s message, adding start item",
+                    full_item.mode.name()), UVM_MEDIUM)
                 message.add_start_msg_item(full_item);
               end else begin
+                `uvm_info(`gfn, $sformatf("rebuilding %s message, adding data block #%0d",
+                    full_item.mode.name(), message.output_msg.size()/16), UVM_MEDIUM)
                 message.add_data_item(full_item);
               end
             end
@@ -501,30 +523,48 @@ virtual task rebuild_message();
         end
       end
 
-      begin
-        // AES indicates when it's done with processing individual blocks but not when it's done
-        // with processing an entire message. To detect the end of a message, the DV environment
-        // does the following:
-        // - It tracks writes to the main control register. If two successfull writes to this
-        //   shadowed register are observed, this marks the start of a new message.
-        // - DV then knows that the last output data retrieved marks the end of the previous
-        //   message.
-        // This works fine except for the very last message before the test ends. To mark the end
-        // of the last message, the `finish_message` variable is used. It gets set by the
-        // `phase_ready_to_end()` function, see below.
-        wait (finish_message)
-        `uvm_info(`gfn, $sformatf("\n\t ----| Finish test received adding message item to mg_fifo"),
-                                  UVM_MEDIUM)
-        `downcast(msg_clone, message.clone());
-        msg_fifo.put(msg_clone);
+      begin : finish_last_message
+        forever begin
+          // AES indicates when it's done with processing individual blocks but not when it's done
+          // with processing an entire message. To detect the end of a message, the DV environment
+          // does the following:
+          // - It tracks writes to the main control register. If two successfull writes to this
+          //   shadowed register are observed, this marks the start of a new message.
+          // - DV then knows that the last output data retrieved marks the end of the previous
+          //   message.
+          // This works fine except for the very last message before the sequence ends. To mark the
+          // end of the last message, the `finish_message` variable is used. It gets set by the
+          // `post_body()` task defined in the base vseq.
+          wait (cfg.finish_message)
+          `uvm_info(`gfn, $sformatf("finish_message received"), UVM_MEDIUM)
+          if (msg_state != MSG_IDLE) begin
+            // The message rebuilding thread isn't idle, i.e., there was some activity since the
+            // last reset.
+            `downcast(msg_clone, message.clone());
+            `uvm_info(`gfn, $sformatf("adding %s message item of size %0d to msg_fifo",
+                msg_clone.aes_mode.name(), msg_clone.output_msg.size()), UVM_MEDIUM)
+            msg_fifo.put(msg_clone);
+            // Reset the message rebuilding thread.
+            reset_rebuilding = 1;
+          end
+          // Increment counter for total number of messages seen. If the message rebuilding thread
+          // is still idle, no message did actually go through the DUT and we can skip incrementing
+          // the counter. The only exception is if all messages were corrupted. The DUT doesn't
+          // produce any output in this case.
+          if ((msg_state != MSG_IDLE) || (cfg.num_messages == cfg.num_corrupt_messages)) begin
+            cfg.num_messages_tot += cfg.num_messages;
+          end
+          wait_fifo_empty();
+          cfg.finish_message = 0;
+        end
       end
 
-      begin: check_for_reset
+      begin: reset_rebuild_messages
         forever begin
-          wait (reset_compare);
-          msg_state = MSG_START;
+          wait (reset_rebuilding);
+          msg_state = MSG_IDLE;
           message = new();
-          reset_compare = 0;
+          reset_rebuilding = 0;
         end
       end
     join
@@ -592,21 +632,9 @@ virtual task rebuild_message();
     end
   endtask
 
+
   virtual function void phase_ready_to_end(uvm_phase phase);
     if (phase.get_name() != "run") return;
-
-    // AES indicates when it's done with processing individual blocks but not when it's done
-    // with processing an entire message. To detect the end of a message, the DV environment
-    // does the following:
-    // - It tracks writes to the main control register. If two successfull writes to this
-    //   shadowed register are observed, this marks the start of a new message.
-    // - DV then knows that the last output data retrieved marks the end of the previous
-    //   message.
-    // This works fine except for the very last message before the test ends. To mark the end
-    // of the last message, and trigger its scoring, the `finish_message` variable is set. It
-    // gets read by the `rebuild_message()` task above.
-    finish_message = 1;
-    `uvm_info(`gfn, $sformatf("Finish message: %b", finish_message), UVM_MEDIUM)
 
     // Don't end the test yet. First, the last message needs to be scored, and all queues and
     // FIFOs need to be emptied.
@@ -646,7 +674,7 @@ virtual task rebuild_message();
     // if split is set before reset make sure to cancel
     input_item.split_item = 0;
     // reset compare task to start
-    reset_compare = 1;
+    reset_rebuilding = 1;
   endfunction
 
 

--- a/hw/ip/aes/dv/env/aes_scoreboard.sv
+++ b/hw/ip/aes/dv/env/aes_scoreboard.sv
@@ -21,10 +21,6 @@ class aes_scoreboard extends cip_base_scoreboard #(
   aes_seq_item key_item;                      // sequence item holding last sideload valid key
   bit          ok_to_fwd          = 0;        // 0: item is not ready to forward
   bit          finish_message     = 0;        // set when test is trying to end
-                                              // - to indicate the last message is finished
-  int          good_cnt           = 0;        // number of good messages
-  int          corrupt_cnt        = 0;        // number of aes_mode errors seen
-  int          skipped_cnt        = 0;        // number of skipped messages
   bit          reset_compare      = 0;        // reset compare task
   bit          exp_clear          = 0;        // if using sideload - we are expecting a clear
   keymgr_pkg::hw_key_req_t sideload_key = 0;  // will hold the key from sideload
@@ -572,25 +568,25 @@ virtual task rebuild_message();
             txt = {txt,
                  $sformatf("\n\t ----| FAILED AT BYTE #%0d \t ACTUAL: 0x%h \t PREDICTED: 0x%h ",
                                   n, msg.output_msg[n], msg.predicted_msg[n])};
-          `uvm_fatal(`gfn, $sformatf(" # %0d  \n\t %s \n", good_cnt, txt))
+          `uvm_fatal(`gfn, $sformatf(" # %0d  \n\t %s \n", cfg.good_cnt, txt))
           end
         end
-        `uvm_info(`gfn, $sformatf("\n\t ----|   MESSAGE #%0d MATCHED  %s  |-----",
-                      good_cnt, msg.aes_mode.name() ),
-                                  UVM_MEDIUM)
-        good_cnt++;
+        `uvm_info(`gfn,
+            $sformatf("\n\t ----|   MESSAGE #%0d MATCHED  %s  |-----",
+                cfg.good_cnt, msg.aes_mode.name()), UVM_MEDIUM)
+        cfg.good_cnt++;
       end else begin
         if (msg.aes_mode == AES_NONE) begin
           `uvm_info(`gfn,
-                    $sformatf("\n\t ----| MESSAGE #%0d HAS ILLEGAL MODE MESSAGE IGNORED     |-----",
-                    good_cnt), UVM_MEDIUM)
-          corrupt_cnt++;
+              $sformatf("\n\t ----| MESSAGE #%0d HAS ILLEGAL MODE MESSAGE IGNORED     |-----",
+                  cfg.good_cnt), UVM_MEDIUM)
+          cfg.corrupt_cnt++;
         end
         if (msg.skip_msg) begin
           `uvm_info(`gfn,
-                 $sformatf("\n\t ----| MESSAGE #%0d was skipped due to start triggered prematurely",
-                  good_cnt), UVM_MEDIUM)
-          skipped_cnt++;
+              $sformatf("\n\t ----| MESSAGE #%0d was skipped due to start triggered prematurely",
+                  cfg.good_cnt), UVM_MEDIUM)
+          cfg.skipped_cnt++;
         end
       end
     end
@@ -642,8 +638,10 @@ virtual task rebuild_message();
     while (item_fifo.try_get(seq_item));
     while (msg_fifo.try_get(msg_item));
 
-    skipped_cnt   = 0;
-    good_cnt      = 0;
+    cfg.num_messages_tot = 0;
+    cfg.good_cnt = 0;
+    cfg.corrupt_cnt = 0;
+    cfg.skipped_cnt = 0;
     cfg.split_cnt = 0;
     // if split is set before reset make sure to cancel
     input_item.split_item = 0;
@@ -652,71 +650,72 @@ virtual task rebuild_message();
   endfunction
 
 
-  function void check_phase(uvm_phase phase);
-    string txt =  "";
-    uvm_report_server rpt_srvr;
+  function string counters2string(string txt);
+      txt = { txt, $sformatf("\n\t ----| Expected:           %0d", cfg.num_messages_tot)};
+      txt = { txt, $sformatf("\n\t ----| Seen:               %0d", cfg.good_cnt)};
+      txt = { txt, $sformatf("\n\t ----| Expected corrupted: %0d", cfg.num_corrupt_messages)};
+      txt = { txt, $sformatf("\n\t ----| Seen corrupted:     %0d", cfg.corrupt_cnt)};
+      txt = { txt, $sformatf("\n\t ----| Skipped:            %0d", cfg.skipped_cnt)};
+      txt = { txt, $sformatf("\n\t ----| Split:              %0d", cfg.split_cnt)};
+      return txt;
+  endfunction
 
+
+  function void check_message_counters();
+    uvm_report_server rpt_srvr;
+    string txt = "";
+    // check that we saw all messages
+    // if there is more than expected check split count
+    if (cfg.good_cnt <
+        (cfg.num_messages_tot - cfg.num_corrupt_messages - cfg.skipped_cnt)) begin
+      rpt_srvr = uvm_report_server::get_server();
+      if (rpt_srvr.get_severity_count(UVM_FATAL)
+           + rpt_srvr.get_severity_count(UVM_ERROR) == 0) begin
+        txt = "\n\t ----| NO FAILURES BUT NUMBER OF EXPECTED MESSAGES DOES NOT MATCH ACTUAL";
+      end else begin
+        txt = "\n\t ----| TEST FAILED";
+      end
+      txt = counters2string(txt);
+      `uvm_fatal(`gfn, $sformatf("%s", txt))
+    end
+    if ((cfg.good_cnt >
+        (cfg.num_messages_tot - cfg.num_corrupt_messages - cfg.skipped_cnt))
+        && (cfg.split_cnt == 0)) begin
+      txt = "\n\t ----| SAW TOO MANY MESSAGES AND NONE WAS SPLIT";
+      txt = counters2string(txt);
+      `uvm_fatal(`gfn, $sformatf("%s", txt))
+    end
+  endfunction
+
+
+  function void check_phase(uvm_phase phase);
     if (cfg.en_scb) begin
       super.check_phase(phase);
       `DV_EOT_PRINT_MAILBOX_CONTENTS(aes_message_item, msg_fifo)
       `DV_EOT_PRINT_MAILBOX_CONTENTS(aes_seq_item, item_fifo)
       `DV_EOT_PRINT_Q_CONTENTS(aes_seq_item, rcv_item_q)
-      // check that we saw all messages
-      // if there is more than expected check split count
-      if (good_cnt <
-             (cfg.num_messages - cfg.num_corrupt_messages - skipped_cnt)) begin
-        rpt_srvr = uvm_report_server::get_server();
-        if (rpt_srvr.get_severity_count(UVM_FATAL)
-             + rpt_srvr.get_severity_count(UVM_ERROR) == 0) begin
-          txt = "\n\t ----| NO FAILURES BUT NUMBER OF EXPECTED MESSAGES DOES NOT MATCH ACTUAL";
-        end else begin
-          txt = "\n\t ----| TEST FAILED";
-        end
-
-        txt = { txt, $sformatf(" \n\t ----| Expected:\t %d", cfg.num_messages)};
-        txt = { txt, $sformatf(" \n\t ----| Seen: \t%d",  good_cnt)};
-        txt = { txt, $sformatf(" \n\t ----| Expected corrupted: \t%d", cfg.num_corrupt_messages)};
-        txt = { txt, $sformatf(" \n\t ----| Seen corrupted: \t%d", corrupt_cnt)};
-        txt = { txt, $sformatf(" \n\t ----| Skipped: \t%d", skipped_cnt)};
-        txt = { txt, $sformatf(" \n\t ----| Split: \t%d", cfg.split_cnt)};
-        `uvm_fatal(`gfn, $sformatf("%s", txt) )
-      end
-      if ((good_cnt > (cfg.num_messages - cfg.num_corrupt_messages - skipped_cnt))
-           && (cfg.split_cnt == 0)) begin
-        txt = " SAW TOO MANY MESSAGES AND NONE WAS SPLIT";
-        txt = { txt, $sformatf(" \n\t ----| Expected:\t %d", cfg.num_messages)};
-        txt = { txt, $sformatf(" \n\t ----| Seen: \t%d",  good_cnt)};
-        txt = { txt, $sformatf(" \n\t ----| Expected corrupted: \t%d", cfg.num_corrupt_messages)};
-        txt = { txt, $sformatf(" \n\t ----| Seen corrupted: \t%d", corrupt_cnt)};
-        txt = { txt, $sformatf(" \n\t ----| Skipped: \t%d", skipped_cnt)};
-        txt = { txt, $sformatf(" \n\t ----| Split: \t%d", cfg.split_cnt)};
-        `uvm_fatal(`gfn, $sformatf("%s", txt))
-      end
+      check_message_counters();
     end
   endfunction
 
 
   function void report_phase(uvm_phase phase);
     uvm_report_server rpt_srvr;
-    string txt="";
+    string txt = "";
 
     super.report_phase(phase);
-    txt = $sformatf("\n\t ----|        TEST FINISHED        |----");
-    txt = { txt, $sformatf("\n\t Saw %d Good messages  ", good_cnt)};
-    txt = { txt, $sformatf("\n\t Skipped %d messages  " , skipped_cnt)};
-    txt = { txt, $sformatf("\n\t Split %d messages  "   , cfg.split_cnt)};
-    txt = { txt, $sformatf("\n\t Expected %d messages  ", cfg.num_messages)};
-    txt = { txt, $sformatf("\n\t Expected %d corrupted ", cfg.num_corrupt_messages)};
+    txt = "\n\t ----|        TEST FINISHED        |----";
+    txt = counters2string(txt);
     rpt_srvr = uvm_report_server::get_server();
-    if (rpt_srvr.get_severity_count(UVM_FATAL)+rpt_srvr.get_severity_count(UVM_ERROR)>0) begin
+    if (rpt_srvr.get_severity_count(UVM_FATAL) + rpt_srvr.get_severity_count(UVM_ERROR) > 0) begin
       `uvm_info(`gfn, $sformatf("%s", cfg.convert2string()), UVM_LOW)
-      txt = { txt,"\n\t---------------------------------------"};
-      txt = { txt,"\n\t----            TEST FAILED        ----"};
-      txt = { txt,"\n\t---------------------------------------"};
+      txt = {txt, "\n\t ---------------------------------------"};
+      txt = {txt, "\n\t ----            TEST FAILED        ----"};
+      txt = {txt, "\n\t ---------------------------------------"};
     end else begin
-      txt = {txt, "\n\t---------------------------------------"};
-      txt = { txt,"\n\t----            TEST PASSED        ----"};
-      txt = { txt,"\n\t---------------------------------------"};
+      txt = {txt, "\n\t ---------------------------------------"};
+      txt = {txt, "\n\t ----            TEST PASSED        ----"};
+      txt = {txt, "\n\t ---------------------------------------"};
     end
     `uvm_info(`gfn, $sformatf("%s", txt), UVM_MEDIUM)
 

--- a/hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
+++ b/hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
@@ -933,6 +933,30 @@ class aes_base_vseq extends cip_base_vseq #(
     end
   endtask // send_msg_queue
 
+  virtual task post_body();
+
+    // AES indicates when it's done with processing individual blocks but not when it's done
+    // with processing an entire message. To detect the end of a message, the DV environment
+    // does the following:
+    // - It tracks writes to the main control register. If two successfull writes to this
+    //   shadowed register are observed, this marks the start of a new message.
+    // - DV then knows that the last output data retrieved marks the end of the previous
+    //   message.
+    // This works fine except for the very last message before the sequence ends. To mark the
+    // end of the last message, and trigger its scoring, the `finish_message` variable is set.
+    // It gets read by the `rebuild_message()` task in the scoreboard.
+    //
+    // Before doing this, wait for the DUT to become idle and final output to be read.
+    `uvm_info(`gfn, "waiting for DUT to become idle and final output to be read", UVM_MEDIUM)
+    csr_spinwait(.ptr(ral.status.idle), .exp_data(1'b1));
+    csr_spinwait(.ptr(ral.status.output_valid), .exp_data(1'b0));
+    `uvm_info(`gfn, "sending finish_message", UVM_MEDIUM)
+    cfg.finish_message = 1;
+
+    super.post_body();
+
+  endtask
+
 
   ///////////////////////////////////////////////////////////
   ///////////////        FUNCTIONS       ////////////////////

--- a/hw/ip/aes/dv/env/seq_lib/aes_stress_all_vseq.sv
+++ b/hw/ip/aes/dv/env/seq_lib/aes_stress_all_vseq.sv
@@ -1,0 +1,39 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// A sequence that runs other sequences one after the other and randomly injects resets between
+// sequences.
+
+class aes_stress_all_vseq extends aes_base_vseq;
+  `uvm_object_utils(aes_stress_all_vseq)
+  `uvm_object_new
+
+  // The sequences that we'll run back-to-back.
+  string vseq_names[$] = {
+    "aes_stress_vseq",
+    "aes_alert_reset_vseq",
+    "aes_reseed_vseq"
+  };
+
+  task body();
+    `uvm_info(`gfn, $sformatf("Running %0d sub-sequences", num_trans), UVM_LOW)
+    for (int i = 0; i < num_trans; i++) begin
+      uvm_sequence   seq;
+      aes_base_vseq  aes_vseq;
+      uint           seq_idx = $urandom_range(0, vseq_names.size() - 1);
+      string         cur_vseq_name = vseq_names[seq_idx];
+
+      seq = create_seq_by_name(cur_vseq_name);
+      `downcast(aes_vseq, seq)
+
+      aes_vseq.do_apply_reset = (do_apply_reset) ? $urandom_range(0, 1) : 0;
+      aes_vseq.set_sequencer(p_sequencer);
+      `uvm_info(`gfn, $sformatf("iteration[%0d]: starting %0s, resetting %0d",
+          i, cur_vseq_name, aes_vseq.do_apply_reset), UVM_LOW)
+      aes_vseq.start(p_sequencer);
+      `uvm_info(`gfn, $sformatf("iteration[%0d]: ending %0s", i, cur_vseq_name), UVM_LOW)
+    end
+  endtask
+
+endclass

--- a/hw/ip/aes/dv/env/seq_lib/aes_vseq_list.sv
+++ b/hw/ip/aes/dv/env/seq_lib/aes_vseq_list.sv
@@ -17,3 +17,4 @@
 `include "aes_ctr_fi_vseq.sv"
 `include "aes_core_fi_vseq.sv"
 `include "aes_readability_vseq.sv"
+`include "aes_stress_all_vseq.sv"

